### PR TITLE
Increase code coverage

### LINKED ISSUE #18: Increase code coverage

### PR DESCRIPTION
**Pull request by eutampieri**
Closes #18

### CODE REVIEW

**Review by eutampieri — COMMENTED** [review]
First review OK

File: src/main/java/eu/eutampieri/catacombs/model/Entity.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+/**
+ * Abstract class for every living game Entity.
+ */
+public abstract class Entity extends GameObject implements LivingCharacter{
+
+    public static final int FACE_RIGHT = 0;
+    public static final int FACE_LEFT = 1;
+    public static final int FACE_DOWN = 2;
+    public static final int FACE_UP = 3;
```
Comment (eutampieri) [inline src/main/java/eu/eutampieri/catacombs/model/Entity.java:14]: Why not creating an enum here?

File: src/main/java/eu/eutampieri/catacombs/model/Entity.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+    protected int hp; //Entity health
+    protected boolean isAlive;
+    protected int width, height; //Entity width and height
+
+    protected TileMap tileMap;
+    protected CollisionBox hitBox; //Entity hit box
+
+    /**
+     * Entity constructor.
```
Comment (eutampieri) [inline src/main/java/eu/eutampieri/catacombs/model/Entity.java:26]: Maybe say that by default Entity is an enemy

**Review by MyK00L — APPROVED** [review]
lgtm

File: .gitignore [hunk 3]
```diff
@@ -3,4 +3,5 @@
 !src/**/build/
 .gradle
 .settings
-/build/*
\ No newline at end of file
+/build/*
+.idea/
```
Comment (MyK00L) [inline .gitignore:7]: I would also add .classpath and .project

File: src/main/java/eu/eutampieri/catacombs/model/CollisionBox.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        this.height = box.height;
+    }
+
+    /**
+     * Method that moves the collision box by specified values.
+     * @param dx Movement made in the X axis
+     * @param dy Movement made in the Y axis
+     */
+    public void move(final int dx, final int dy) {
+        if (posX + dx > 0 && posY + dy > 0) {
```
Comment (MyK00L) [inline src/main/java/eu/eutampieri/catacombs/model/CollisionBox.java:48]: why this check?

File: src/main/java/eu/eutampieri/catacombs/model/Entity.java [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+     */
+    protected Face face;
+    /**
+     * Entity current health.
+     */
+    protected int hp;
+    /**
+     * Entity dimensions.
+     */
+    protected int width, height; //Entity width and height
```
Comment (MyK00L) [inline src/main/java/eu/eutampieri/catacombs/model/Entity.java:25]: Are these not already in hitBox?

File: src/main/java/eu/eutampieri/catacombs/model/Face.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package eu.eutampieri.catacombs.model;
+
+public enum Face {
```
Comment (MyK00L) [inline src/main/java/eu/eutampieri/catacombs/model/Face.java:3]: I would remove reduntant FACE_ from every name and maybe call the whole enum Directon instead of Face.

File: src/main/java/eu/eutampieri/catacombs/model/map/TileMapImpl.java [hunk 7]
```diff
@@ -1,25 +1,28 @@
...
+    public int width() {
+        return map[0].length;
+    }
+
+    @Override
+    public Tile at(final int x, final int y) {
+        return map[y][x];
+    }
+
+    public Tile[][] getMap() {
```
Comment (MyK00L) [inline src/main/java/eu/eutampieri/catacombs/model/map/TileMapImpl.java:25]: If this is needed it should probabilly be a method of the interface.

File: src/main/java/eu/eutampieri/catacombs/ui/Game.java [hunk 8]
```diff
@@ -11,244 +11,251 @@
...
+    private static MainWindow mainFrame;
+    private static GameConfiguration gameConfiguration;
+
+    private BufferStrategy bs;
+    private GraphicsConfiguration gc;
+    private VolatileImage vImage;
+    private int framesThisSecond;
+    private boolean running = true;
+    private Graphics2D graphics;
+    private int fps;
```
Comment (MyK00L) [inline src/main/java/eu/eutampieri/catacombs/ui/Game.java:39]: We already have fps in GameConfiguration, no?

File: src/main/java/eu/eutampieri/catacombs/ui/MenuState.java [hunk 9]
```diff
@@ -7,68 +7,72 @@
...
-		g2.fillOval(x3, y, 20, 20);
-		
-		
-		
-	}
-	
+public final class MenuState extends State {
+
+    private static final int START_GAME = 1;
+    private static final int QUIT_GAME = 2;
```
Comment (MyK00L) [inline src/main/java/eu/eutampieri/catacombs/ui/MenuState.java:13]: use an enum instead

File: src/main/java/eu/eutampieri/catacombs/ui/MenuState.java [hunk 10]
```diff
@@ -7,68 +7,72 @@
...
+    private static final int TITLE_FONT_SIZE = 50;
+    private static final int DEFAULT_FONT_SIZE = 40;
```
Comment (MyK00L) [inline src/main/java/eu/eutampieri/catacombs/ui/MenuState.java:15]: Maybe we should make all these default options global, so that they're not all over the place

File: src/main/java/eu/eutampieri/catacombs/ui/Game.java [hunk 11]
```diff
@@ -11,244 +11,336 @@
...
-	public void removeKeyAdapter(KeyAdapter e) {
-		mainFrame.getCanvas().removeKeyListener(e);
-		mainFrame.getFrame().removeKeyListener(e);
-	}	
-	
+
+    protected static final double FPS_CONST = 1_000_000_000d;
+    protected static final float DELTA_CONST = 1_000_000_000.0f;
+    protected static final float DELTA_MIN = 0.016f;
+    protected static final float THREAD_SLEEP_CONST = 1_000_000f;
```
Comment (eutampieri) [inline src/main/java/eu/eutampieri/catacombs/ui/Game.java:24]: These names suck!

File: src/main/java/eu/eutampieri/catacombs/ui/Game.java [hunk 12]
```diff
@@ -11,244 +11,336 @@
...
+    }
+
+    /**
+     *
+     * @param config
+     *              the style choose for the frame
+     */
+
+    // TODO Is this warning really necessary? Isn't it bad design?
+    @SuppressFBWarnings("ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD")
```
Comment (eutampieri) [inline src/main/java/eu/eutampieri/catacombs/ui/Game.java:106]: It was bad design and it's been refactored so no, this is no longer needed

File: src/main/java/eu/eutampieri/catacombs/ui/MenuState.java (hunk 10 above)
Comment (eutampieri) [reply src/main/java/eu/eutampieri/catacombs/ui/MenuState.java:15]: We should definitely investigate this, maybe a static class per package?

File: src/main/java/eu/eutampieri/catacombs/ui/State.java [hunk 13]
```diff
@@ -3,18 +3,49 @@
...
+     * @param delta
+     *              gap time from the previous render
+     */
+
+    public abstract void update(float delta);
+
+    /**
+     *
+     * @param g2
+     *              use to generate graphics
```
Comment (eutampieri) [inline src/main/java/eu/eutampieri/catacombs/ui/State.java:37]: This should be placed above `@param`

File: src/main/java/eu/eutampieri/catacombs/ui/State.java [hunk 14]
```diff
@@ -3,18 +3,49 @@
...
+     */
+
+    public State(final DungeonGame game) {
+        this.game = game;
+    }
+
+    /**
+     *
+     * @param delta
+     *              gap time from the previous render
```
Comment (eutampieri) [inline src/main/java/eu/eutampieri/catacombs/ui/State.java:29]: Same as below

File: src/main/java/eu/eutampieri/catacombs/ui/State.java [hunk 15]
```diff
@@ -3,18 +3,49 @@
...
+    /**
+     *
+     * var used in the constructor
+     *
+     */
+
+    private final DungeonGame game;
+
+    /**
+     *
```
Comment (eutampieri) [inline src/main/java/eu/eutampieri/catacombs/ui/State.java:16]: Too many empty lines

File: src/main/java/eu/eutampieri/catacombs/ui/State.java [hunk 16]
```diff
@@ -3,18 +3,49 @@
...
+     *
+     * @param game
+     *             manages the states
```
Comment (eutampieri) [inline src/main/java/eu/eutampieri/catacombs/ui/State.java:19]: The first line should contain the function brief description

File: src/main/java/eu/eutampieri/catacombs/ui/input/MouseManager.java [hunk 17]
```diff
@@ -3,54 +3,55 @@
...
+
+    @Override
+    public void mouseReleased(final MouseEvent e) {
+        if (e.getButton() == MouseEvent.BUTTON1) {
+            this.leftPress = false;
+        }
+        if (e.getButton() == MouseEvent.BUTTON3) {
+            this.rightPress = false;
+        }
+    }
```
Comment (eutampieri) [inline src/main/java/eu/eutampieri/catacombs/ui/input/MouseManager.java:47]: Beware! After this change (that I did) you will have to detect clicks before the mouse buttons are released otherwise you'll loose the event

File: src/test/java/eu/eutampieri/catacombs/tests/CharactersTests.java [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+
+    }
+
+    /*@Test
+    void testBoxSize() {
+        assertEquals(BAT.getHitBox().getHeight(), 32);
+        assertEquals(BAT.getHitBox().getWidth(), 32);
+        assertEquals(SLIME.getHitBox().getHeight(), 32);
+        assertEquals(SLIME.getHitBox().getWidth(), 32);
+    }*/
```
Comment (eutampieri) [inline src/test/java/eu/eutampieri/catacombs/tests/CharactersTests.java:71]: Maybe we should keep this test?

File: src/test/java/eu/eutampieri/catacombs/tests/CharactersTests.java [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+
+    @Test
+    void testSlimeFollowBat() {
```
Comment (eutampieri) [inline src/test/java/eu/eutampieri/catacombs/tests/CharactersTests.java:74]: This test fails. I think @jackmounts is having a look. Maybe @MyK00L could have a look too

File: src/test/java/eu/eutampieri/catacombs/tests/CollisionBoxTests.java [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Test
+    void testOverlapsWithOverlappingBox() {
+        final CollisionBox overlapping = new CollisionBox(MY_BOX);
+        overlapping.move(MY_BOX.getWidth() / 2, MY_BOX.getHeight() / 2);
+        assertTrue(MY_BOX.overlaps(overlapping));
+    }
+
+    @Test
+    void testDoesntOverlapWithTouchingBox() {
```
Comment (eutampieri) [inline src/test/java/eu/eutampieri/catacombs/tests/CollisionBoxTests.java:34]: According to @jackmounts this is the expected behavior

File: src/main/java/eu/eutampieri/catacombs/model/map/TileMapFactoryImpl.java [hunk 21]
```diff
@@ -4,210 +4,268 @@
...
+     * @param nRooms      number of rooms to generate
+     * @param minRoomSide minimum side length of a room
+     * @param maxRoomSide maximum side length of a room
+     * @param minRoomDist minimum distance between two rooms' centers
+     * @param maxRoomDist maximum distance with the closest room's center for each
+     *                    room center
+     * @return A Tilemap with nRooms square rooms connected by corridors in a tree,
+     *         plus some random corridors minRoomDist > maxRoomSide is recommended
+     */
+    private TileMap normal(final int nRooms, final int minRoomSide, final int maxRoomSide, final int minRoomDist,
```
Comment (eutampieri) [inline src/main/java/eu/eutampieri/catacombs/model/map/TileMapFactoryImpl.java:123]: This function is probably too big

**Review by eutampieri — COMMENTED** [review]
There are changes to be made

File: src/main/java/eu/eutampieri/catacombs/model/GameObject.java [hunk 22]
```diff
@@ -1,86 +1,133 @@
...
-	public void setPosY(int posY) {
-		this.posY = posY;
-	}
+    /**
+     * Getter for object Y position.
+     * @return Object Y position
+     */
+    public int getPosY() {
+        return posY;
+    }
```
[the remark this answers is not in the document]
Comment (eutampieri) [reply src/main/java/eu/eutampieri/catacombs/model/GameObject.java:65]: Ok these are spaces, noice

**Comment by eutampieri** [comment]
I think that this PR can be merged as long as unit tests pass

File: src/test/java/eu/eutampieri/catacombs/tests/CollisionBoxTests.java (hunk 20 above)
Comment (MyK00L) [reply src/test/java/eu/eutampieri/catacombs/tests/CollisionBoxTests.java:34]: It makes sense, if the area of overlap is 0, you can say there is no overlap, this literal edge case shouldn't really matter in the end anyway.

File: build.gradle.kts [hunk 23]
```diff
@@ -81,4 +88,20 @@ tasks.jacocoTestCoverageVerification {
...
+    ruleSetFiles(file("$projectDir/config/pmd/pmd.xml"))
+}
+
+tasks.spotbugsMain {
+    reports.create("html") {
+        isEnabled = true
+        setStylesheet("fancy-hist.xsl")
+    }
+}
+
```
Comment (MyK00L) [inline build.gradle.kts:107]: `mainClassName = "eu.eutampieri.catacombs.game.Catcombs"` generates a deprecated warning, we should consider changing it with whatever is the non deprecated way of doing this.

Comment (eutampieri) [reply build.gradle.kts:107]: It is but it comes from Danilo Pianini's example.
The warning doesn't stop the checks from succeding I think

**Comment by eutampieri** [comment]
Did someone write a test that opens a GUI? xD

**Comment by MyK00L** [comment]
> Did someone write a test that opens a GUI? xD

I wrote a test for Game and it calls run, so it opens a GUI for a fraction of a second, but it should close it right away.

**Comment by eutampieri** [comment]
@MyK00L let's see if this breaks the CI pipeline!

**Comment by eutampieri** [comment]
@MyK00L  what shall I say...
```java
    java.awt.HeadlessException at DungeonGameTest.java:22
```

**Review by MyK00L — APPROVED** [review]

**Comment by eutampieri** [comment]
What shall I say? I'm a simple man: I see a green check, I merge! :tada:
